Intersection breaks generic key field inference

<!-- 🚨 STOP 🚨 𝗦𝗧𝗢𝗣 🚨 𝑺𝑻𝑶𝑷 🚨

Half of all issues filed here are duplicates, answered in the FAQ, or not appropriate for the bug tracker. Even if you think you've found a *bug*, please read the FAQ first, especially the Common "Bugs" That Aren't Bugs section!

Please help us by doing the following steps before logging an issue:
  * Search: https://github.com/Microsoft/TypeScript/search?type=Issues
  * Read the FAQ: https://github.com/Microsoft/TypeScript/wiki/FAQ

Please fill in the *entire* template below.
-->

<!-- Please try to reproduce the issue with `typescript@next`. It may have already been fixed. -->
**TypeScript Version:**  3.1.0-dev.20180810

**Code**

```ts
const f = <T extends { [K0 in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
{
  const s: string = t[k];
};
```

**Expected behavior:**  Successful compilation

**Actual behavior:**  I get the following error:
```
error TS2322: Type 'T[K]' is not assignable to type 'string'.

     const s: string = t[k];
           ~
```

**Playground Link:** [Link](https://www.typescriptlang.org/play/#src=const%20f%20%3D%20<T%20extends%20%7B%20%5BK0%20in%20K%5D%3A%20string%3B%20%7D%20%26%20%7B%20cool%3A%20string%3B%20%7D%2C%20K%20extends%20keyof%20T>(t%3A%20T%2C%20k%3A%20K)%20%3D>%0D%0A%7B%0D%0A%20%20const%20s%3A%20string%20%3D%20t%5Bk%5D%3B%0D%0A%7D%3B%0D%0A)

**Additional Notes:**  This compiles successfully ([link](https://www.typescriptlang.org/play/#src=const%20f%20%3D%20<T%20extends%20%7B%20%5BK0%20in%20K%5D%3A%20string%3B%20%7D%20%26%20%7B%20cool%3A%20string%3B%20%7D%2C%20K%20extends%20keyof%20T>(t%3A%20T%2C%20k%3A%20K)%20%3D>%0D%0A%7B%0D%0A%20%20const%20s%3A%20string%20%3D%20t%5B'cool'%5D%3B%0D%0A%7D%3B%0D%0A)):
```ts
const f = <T extends { [K0 in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
{
  const s: string = t['cool'];
};
```
as does this ([link](https://www.typescriptlang.org/play/#src=const%20f%20%3D%20<T%20extends%20%7B%20%5BK0%20in%20K%5D%3A%20string%3B%20%7D%2C%20K%20extends%20keyof%20T>(t%3A%20T%2C%20k%3A%20K)%20%3D>%0D%0A%7B%0D%0A%20%20const%20s%3A%20string%20%3D%20t%5Bk%5D%3B%0D%0A%7D%3B%0D%0A)):
```ts
const f = <T extends { [K0 in K]: string; }, K extends keyof T>(t: T, k: K) =>
{
  const s: string = t[k];
};
```

--- a/src/compiler/checker.ts
+++ b/src/compiler/checker.ts
@@ -6899,10 +6899,15 @@ namespace ts {
         }
 
         function getConstraintOfIndexedAccess(type: IndexedAccessType) {
-            const objectType = getBaseConstraintOfType(type.objectType) || type.objectType;
-            const indexType = getBaseConstraintOfType(type.indexType) || type.indexType;
-            const constraint = !isGenericObjectType(objectType) && !isGenericIndexType(indexType) ? getIndexedAccessType(objectType, indexType, /*accessNode*/ undefined, errorType) : undefined;
-            return constraint && constraint !== errorType ? constraint : undefined;
+            const objectType = getConstraintOfType(type.objectType) || type.objectType;
+            if (objectType !== type.objectType) {
+                const constraint = getIndexedAccessType(objectType, type.indexType, /*accessNode*/ undefined, errorType);
+                if (constraint && constraint !== errorType) {
+                    return constraint;
+                }
+            }
+            const baseConstraint = getBaseConstraintOfType(type);
+            return baseConstraint && baseConstraint !== type ? baseConstraint : undefined;
         }
 
         function getDefaultConstraintOfConditionalType(type: ConditionalType) {
@@ -7073,9 +7078,6 @@ namespace ts {
                 }
                 if (t.flags & TypeFlags.Substitution) {
                     return getBaseConstraint((<SubstitutionType>t).substitute);
-                }
-                if (isGenericMappedType(t)) {
-                    return emptyObjectType;
                 }
                 return t;
             }
@@ -11642,12 +11644,13 @@ namespace ts {
                     }
                 }
                 else if (target.flags & TypeFlags.IndexedAccess) {
-                    // A type S is related to a type T[K] if S is related to C, where C is the
-                    // constraint of T[K]
-                    const constraint = getConstraintForRelation(target);
-                    if (constraint) {
-                        if (result = isRelatedTo(source, constraint, reportErrors)) {
-                            return result;
+                    // A type S is related to a type T[K] if S is related to C, where C is the base constraint of T[K]
+                    if (relation !== identityRelation) {
+                        const constraint = getBaseConstraintOfType(target);
+                        if (constraint && constraint !== target) {
+                            if (result = isRelatedTo(source, constraint, reportErrors)) {
+                                return result;
+                            }
                         }
                     }
                 }

--- a/tests/baselines/reference/keyofAndIndexedAccess.js
+++ b/tests/baselines/reference/keyofAndIndexedAccess.js
@@ -650,6 +650,20 @@ function ff2<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: 
     return d[k2];
 }
 
+// Repro from #26409
+
+const cf1 = <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
+{
+    const s: string = t[k];
+    t.cool;
+};
+
+const cf2 = <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) =>
+{
+    const s: string = t[k];
+    t.cool;
+};
+
 
 //// [keyofAndIndexedAccess.js]
 var __extends = (this && this.__extends) || (function () {
@@ -1078,6 +1092,15 @@ function ff2(dd, k1, k2) {
     var d = dd[k1];
     return d[k2];
 }
+// Repro from #26409
+var cf1 = function (t, k) {
+    var s = t[k];
+    t.cool;
+};
+var cf2 = function (t, k) {
+    var s = t[k];
+    t.cool;
+};
 
 
 //// [keyofAndIndexedAccess.d.ts]
@@ -1413,3 +1436,7 @@ declare type DictDict<V extends string, T extends string> = {
 };
 declare function ff1<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: T): number;
 declare function ff2<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: T): number;
+declare const cf1: <T extends { [P in K]: string; } & {
+    cool: string;
+}, K extends keyof T>(t: T, k: K) => void;
+declare const cf2: <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) => void;

--- a/tests/baselines/reference/keyofAndIndexedAccess.symbols
+++ b/tests/baselines/reference/keyofAndIndexedAccess.symbols
@@ -2319,3 +2319,54 @@ function ff2<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: 
 >k2 : Symbol(k2, Decl(keyofAndIndexedAccess.ts, 646, 75))
 }
 
+// Repro from #26409
+
+const cf1 = <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
+>cf1 : Symbol(cf1, Decl(keyofAndIndexedAccess.ts, 653, 5))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 653, 13))
+>P : Symbol(P, Decl(keyofAndIndexedAccess.ts, 653, 26))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 653, 65))
+>cool : Symbol(cool, Decl(keyofAndIndexedAccess.ts, 653, 48))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 653, 65))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 653, 13))
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 653, 85))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 653, 13))
+>k : Symbol(k, Decl(keyofAndIndexedAccess.ts, 653, 90))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 653, 65))
+{
+    const s: string = t[k];
+>s : Symbol(s, Decl(keyofAndIndexedAccess.ts, 655, 9))
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 653, 85))
+>k : Symbol(k, Decl(keyofAndIndexedAccess.ts, 653, 90))
+
+    t.cool;
+>t.cool : Symbol(cool, Decl(keyofAndIndexedAccess.ts, 653, 48))
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 653, 85))
+>cool : Symbol(cool, Decl(keyofAndIndexedAccess.ts, 653, 48))
+
+};
+
+const cf2 = <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) =>
+>cf2 : Symbol(cf2, Decl(keyofAndIndexedAccess.ts, 659, 5))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 659, 13))
+>P : Symbol(P, Decl(keyofAndIndexedAccess.ts, 659, 26))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 659, 54))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 659, 54))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 659, 13))
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 659, 74))
+>T : Symbol(T, Decl(keyofAndIndexedAccess.ts, 659, 13))
+>k : Symbol(k, Decl(keyofAndIndexedAccess.ts, 659, 79))
+>K : Symbol(K, Decl(keyofAndIndexedAccess.ts, 659, 54))
+{
+    const s: string = t[k];
+>s : Symbol(s, Decl(keyofAndIndexedAccess.ts, 661, 9))
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 659, 74))
+>k : Symbol(k, Decl(keyofAndIndexedAccess.ts, 659, 79))
+
+    t.cool;
+>t.cool : Symbol(cool)
+>t : Symbol(t, Decl(keyofAndIndexedAccess.ts, 659, 74))
+>cool : Symbol(cool)
+
+};
+

--- a/tests/baselines/reference/keyofAndIndexedAccess.types
+++ b/tests/baselines/reference/keyofAndIndexedAccess.types
@@ -2184,3 +2184,44 @@ function ff2<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: 
 >k2 : T
 }
 
+// Repro from #26409
+
+const cf1 = <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
+>cf1 : <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) => void
+><T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>{    const s: string = t[k];    t.cool;} : <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) => void
+>cool : string
+>t : T
+>k : K
+{
+    const s: string = t[k];
+>s : string
+>t[k] : T[K]
+>t : T
+>k : K
+
+    t.cool;
+>t.cool : string
+>t : T
+>cool : string
+
+};
+
+const cf2 = <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) =>
+>cf2 : <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) => void
+><T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) =>{    const s: string = t[k];    t.cool;} : <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) => void
+>t : T
+>k : K
+{
+    const s: string = t[k];
+>s : string
+>t[k] : T[K]
+>t : T
+>k : K
+
+    t.cool;
+>t.cool : string
+>t : T
+>cool : string
+
+};
+

--- a/tests/cases/conformance/types/keyof/keyofAndIndexedAccess.ts
+++ b/tests/cases/conformance/types/keyof/keyofAndIndexedAccess.ts
@@ -651,3 +651,17 @@ function ff2<V extends string, T extends string>(dd: DictDict<V, T>, k1: V, k2: 
     const d: Dict<T> = dd[k1];
     return d[k2];
 }
+
+// Repro from #26409
+
+const cf1 = <T extends { [P in K]: string; } & { cool: string; }, K extends keyof T>(t: T, k: K) =>
+{
+    const s: string = t[k];
+    t.cool;
+};
+
+const cf2 = <T extends { [P in K | "cool"]: string; }, K extends keyof T>(t: T, k: K) =>
+{
+    const s: string = t[k];
+    t.cool;
+};

